man/rcvstore.man
.TH RCVSTORE %manext1% 2013-02-19 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
rcvstore \- asynchronously incorporate mail into an nmh folder
.SH SYNOPSIS
.HP 5
.na
.B %nmhlibexecdir%/rcvstore
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RB [ \-create " | " \-nocreate ]
.RB [ \-unseen " | " \-nounseen ]
.RB [ \-zero " | " \-nozero ]
.RB [ \-sequence
.I name
\&...]
.RB [ \-public " | " \-nopublic ]
.ad
.SH DESCRIPTION
.B rcvstore
incorporates a message from the standard input into an
.B nmh
folder.  This command is typically used in conjunction with
mail filtering programs such as
.B slocal
and
.BR procmail ,
to filter your mail into different folders.
.PP
You may specify which folder to use with
.IR +folder .
If no folder is specified,
.B rcvstore
will use the folder given by a non-empty \*(lqInbox:\*(rq entry in
the user's profile, else it will use the folder named \*(lqinbox\*(rq.
.PP
If the switch
.B \-create
is given (it is the default) and if the specified
(or default) folder does not exist, then it will be created.  You may
disable this with the
.B \-nocreate
option.  In this case
.B rcvstore
will exit if the specified folder does not exist.
.PP
When the new message is incorporated into the folder, it is assigned
the next highest number for that folder.
.PP
.B rcvstore
will incorporate anything except zero length messages into the user's
.B nmh
folder.  It will not change the message in any way.
.PP
If the user's profile contains a
.RI \*(lq "Msg\-Protect: nnn" \*(rq
entry, it will be used as the mode to use, in octal, on the newly created
message, otherwise the
.B nmh
default mode of 0600, see
.IR chmod (1),
will be used (on filesystems that support it).
For all subsequent operations on this message, this initially assigned
mode will be preserved.
.PP
If the switch
.B \-unseen
is given (it is on by default), and if the profile entry
\*(lqUnseen\-Sequence\*(rq is present and non-empty, then
.B rcvstore
will add the newly incorporated message to each sequence
named by this profile entry.  You may use the switch
.B \-nounseen
to disable this.  These sequences will not be zeroed by
.B rcvstore
prior to adding the new message.
.PP
Furthermore, the incoming messages may be added to additional sequences
as they arrive by the use of the
.B \-sequence
switch.  As with the commands
.B pick
and
.BR mark ,
you may also use the switches
.B \-zero
and
.B \-nozero
to specify whether to zero old sequences or not.
Similarly, use of the
.B \-public
and
.B \-nopublic
switches may be used to force these sequences to be public or
private sequences.
.SS Locking and sequences
If you use the \*(lqUnseen-Sequence\*(rq profile entry, or if you
use the
.B \-sequence
argument, 
.B rcvstore
might try to update its sequence state while another
.B nmh
process is also trying to do so.  This can cause one of those updates
to be lost.  To avoid this (just between asynchronous invocations of
.BR rcvstore ),
always use it with an external locking mechanism.  A
.IR procmailrc (5)
local lockfile would work, or if not using procmail, then consider using
.IR dotlockfile (1)
or
.IR flock (1).
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Folder\-Protect:~^To set mode when creating a new folder
^Inbox:~^To find the default inbox
^Msg\-Protect:~^To set mode when creating a new message
^Unseen\-Sequence:~^To name sequences denoting unseen messages
.fi
.SH "SEE ALSO"
.IR mh\-sequence (5),
.IR procmailex (5),
.IR procmailrc (5),
.IR rcvdist (1),
.IR rcvpack (1),
.IR rcvtty (1)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to \*(lqInbox\*(rq profile entry"
.RB ` \-create '
.RB ` \-unseen '
.RB ` \-nozero '
.fi
.SH CONTEXT
No context changes will be attempted, with the exception of
sequence manipulation.


NEWS
#
# NEWS - Release notes for nmh 1.8
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly five years since the last release of nmh, and there have
been a number of significant changes since that last release.  Long-time
MH and nmh users should read carefully the NOTABLE CHANGES section, as
there are some significant changes to nmh behavior.  Otherwise, please
see the README and INSTALL files for help on getting started with nmh.

This release is dedicated to Norman Z. Shapiro, co-designer of the MH
Message Handling System.  MH is the predecessor of nmh.  Norm was an
active supporter of nmh development until he passed away in October of
2021.  We are most grateful to Norm for his stewardship of MH and nmh.
    https://en.wikipedia.org/wiki/Norman_Shapiro

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

- Support for Content-MD5 header fields, MIME content cache functionality,
  and the message/partial MIME type have been removed.
- Gmail OAuth2/XOAUTH support for desktop applications has been effectively
  dropped, so nmh no longer supports it.  nmh support for Gmail API access
  is experimental, please post to nmh-workers@nongnu.org if you'd like to
  help with test and development.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.  Because encoding can wrap
  long lines, the use of a paragraph formatter has been removed from
  mhn.defaults.

------------
NEW FEATURES
------------

- The default editor has been changed from 'vi' to 'prompter', to align with
  historical practice and reduce packaging dependencies on external programs.
- A new -checkbase64 switch has been added to mhfixmsg(1).
- inc(1)/msgchk(1) now support STARTTLS for the POP protocol.
- All TLS-supported protocols now will send the SNI (server name indicator)
  TLS extension.
- A new mh-format function %(ordinal) has been implemented to output the
  appropriate ordinal suffix for numbers.  Example: "%(num 22)%(ordinal)"
  will output "22nd".
- show and mhl now decode more addresses in header fields.
- Added warning from all programs that read the profile if the profile
  contains a post entry, which is ignored, but does not contain a
  postproc entry.  In other words, if you get this warning and want
  to suppress it, your options include:
  1) Remove your post profile entry.
  2) Make your post profile entry a comment by prepending it with the #:
     comment indicator.
  3) Add a postproc entry that points to the post that you use.  That can
     be viewed with "mhparam postproc".
- scan(1) -file argument can be a Maildir directory.
- Updated mhn.defaults to prefer mpv(1) over xv(1) and replace mpeg_play(1),
  and to use it for all video types, not just video/mpeg.  And prefer all
  other searched-for pdf viewers over acroread(1).
- Added mhshow-suffix-video.mp4 to mhn.defaults, for use by mhshow(1) and
  send(1).
- Removed support from mhn.defaults for application/x-ivs and text/richtext.
- Changed interpretation of argument to mhfixmsg(1) -decodeheaderfieldbodies
  switch to specify character set of the decoded field bodies.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.
- install-mh(1) now enables the mh-draft(5) draft folder facility.

-----------------
OBSOLETE FEATURES
-----------------

- The generation and verification of a Content-MD5 field has been removed
  without deprecation.  The related -check and -nocheck options now error.
- The MIME content cache functionality has been mostly non-functional since
  the development on nmh, and consequently all of the content caching code
  and related switches (-cache/-rcache/-wcache) have been removed.
- Support for generating and reassembling message/partial messages has been
  removed; it seems that this has been broken since 1.5 and there is very
  little support across MUAs.
- Marked Gmail OAuth2/XOAUTH support as being unsupported.
- Support for the MHPDEBUG environment variable was removed.  It was
  deprecated in nmh 1.7.  The pick(1) -debug switch replaced it.
- The 'libdir' mhparam(1) component was removed.  It was deprecated in
  nmh 1.7, when it was replaced by a new 'libexecdir' component.

---------
BUG FIXES
---------

- Fixed bcc to work with sendmail/pipe, and better documented that dcc
  doesn't work with it [Bug 55700].
- An -attendee switch has been added to mhical(1), for use when more than one
  (or zero) attendees match a user's mailbox.
- Fixed inc(1) and %(me) function escape to not obey Local-Mailbox profile
  component.
- Fixed source charset in mhfixmsg textcharset verbose output.
- Fixed mhfixmsg charset determination of content added with -reformat.
- Fixed file descriptor leak in mhfixmsg when run on multiple input files.
- Fixed mhfixmsg(1) -decodeheaderfilebodies to support mixed encoded/undecoded.
- Fixed memory corruption in post(1) and inc(1) when using XOAUTH2,
  with 4 or more entries in the oauth-authservice file.
- Added alias expansion to From: address for use by sendfrom.
- Removed extra space added before header field bodies by dist(1) to $mhdraft.
- Fixed display of iCalendar object with multiple VEVENTS using a VTIMEZONE.
- Fixed allowable encodings with MIME message types to get closer to RFC 2046.
- Detect other files regardless of backup prefix [Bug #49476].
- Copy if hard link by refile(1) or send(1) fails with EACESS [Bug #56575].

----------------------------------------------------------------------------

#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time

uip/rcvstore.c
/* rcvstore.c -- asynchronously add mail to a folder
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sbr/m_gmprot.h"
#include "sbr/getarguments.h"
#include "sbr/seq_setunseen.h"
#include "sbr/seq_save.h"
#include "sbr/smatch.h"
#include "sbr/cpydata.h"
#include "sbr/getfolder.h"
#include "sbr/folder_read.h"
#include "sbr/folder_free.h"
#include "sbr/folder_addmsg.h"
#include "sbr/context_save.h"
#include "sbr/context_find.h"
#include "sbr/ambigsw.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/seq_add.h"
#include "sbr/error.h"
#include <fcntl.h>
#include <signal.h>
#include "sbr/signals.h"
#include "sbr/mts.h"
#include "sbr/done.h"
#include "sbr/utils.h"
#include "sbr/m_maildir.h"
#include "sbr/m_mktemp.h"
#include "sbr/makedir.h"
#include "sbr/globals.h"

#define RCVSTORE_SWITCHES \
    X("create", 0, CRETSW) \
    X("nocreate", 0, NCRETSW) \
    X("unseen", 0, UNSEENSW) \
    X("nounseen", 0, NUNSEENSW) \
    X("public", 0, PUBSW) \
    X("nopublic", 0, NPUBSW) \
    X("zero", 0, ZEROSW) \
    X("nozero", 0, NZEROSW) \
    X("sequence name", 0, SEQSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(RCVSTORE);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(RCVSTORE, switches);
#undef X


/*
 * name of temporary file to store incoming message
 */
static char *tmpfilenam = NULL;

static void unlink_done(int) NORETURN;

int
main (int argc, char **argv)
{
    int publicsw = -1;
    bool zerosw = false;
    bool create = true;
    bool unseensw = true;
    int fd, msgnum;
    size_t seqp = 0;
    char *cp, *maildir, *folder = NULL, buf[BUFSIZ];
    char **argp, **arguments;
    svector_t seqs = svector_create (0);
    struct msgs *mp;
    struct stat st;

    if (nmh_init(argv[0], true, false)) { return 1; }

    set_done(unlink_done);

    mts_init ();
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /* parse arguments */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		die("-%s unknown", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [switches]",
			  invo_name);
		print_help (buf, switches, 1);
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		done (0);

	    case SEQSW: 
		if (!(cp = *argp++) || *cp == '-')
		    die("missing argument name to %s", argp[-2]);

		svector_push_back (seqs, cp);
		seqp++;
		continue;

	    case UNSEENSW:
		unseensw = true;
		continue;
	    case NUNSEENSW:
		unseensw = false;
		continue;

	    case PUBSW: 
		publicsw = 1;
		continue;
	    case NPUBSW: 
		publicsw = 0;
		continue;

	    case ZEROSW: 
		zerosw = true;
		continue;
	    case NZEROSW: 
		zerosw = false;
		continue;

	    case CRETSW: 
		create = true;
		continue;
	    case NCRETSW: 
		create = false;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		die("only one folder at a time!");
            folder = pluspath (cp);
	} else {
	    die("usage: %s [+folder] [switches]", invo_name);
	}
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /* if no folder is given, use default folder */
    if (!folder)
	folder = getfolder (0);
    maildir = m_maildir (folder);

    /* check if folder exists */
    if (stat (maildir, &st) == -1) {
	if (errno != ENOENT)
	    adios (maildir, "error on folder");
	if (!create)
	    die("folder %s doesn't exist", maildir);
	if (!makedir (maildir))
	    die("unable to create folder %s", maildir);
    }

    if (chdir (maildir) == -1)
	adios (maildir, "unable to change directory to");

    /* ignore a few signals */
    SIGNAL (SIGHUP, SIG_IGN);
    SIGNAL (SIGINT, SIG_IGN);
    SIGNAL (SIGQUIT, SIG_IGN);
    SIGNAL (SIGTERM, SIG_IGN);

    /* create a temporary file */
    tmpfilenam = m_mktemp (invo_name, &fd, NULL);
    if (tmpfilenam == NULL) {
	die("unable to create temporary file in %s", get_temp_dir());
    }
    chmod (tmpfilenam, m_gmprot());

    /* copy the message from stdin into temp file */
    cpydata (fileno (stdin), fd, "standard input", tmpfilenam);

    if (fstat (fd, &st) == -1) {
	(void) m_unlink (tmpfilenam);
	adios (tmpfilenam, "unable to fstat");
    }
    if (close (fd) == -1)
	adios (tmpfilenam, "error closing");

    /* don't add file if it is empty */
    if (st.st_size == 0) {
	(void) m_unlink (tmpfilenam);
	inform("empty file");
	done (0);
    }

    /*
     * read folder and create message structure
     */
    if (!(mp = folder_read (folder, 1)))
	die("unable to read folder %s", folder);

    /*
     * Link message into folder, and possibly add
     * to the Unseen-Sequence's.
     */
    if ((msgnum = folder_addmsg (&mp, tmpfilenam, 0, unseensw, 0, 0, NULL)) == -1)
	done (1);

    /*
     * Add the message to any extra sequences
     * that have been specified.
     */
    if (seqp) {
	/* The only reason that seqp was checked to be non-zero is in
	   case a -nosequence switch is added. */
	for (seqp = 0; seqp < svector_size (seqs); seqp++) {
	    if (!seq_addmsg (mp, svector_at (seqs, seqp), msgnum, publicsw,
			     zerosw))
		done (1);
	}
    }

    svector_free (seqs);
    seq_setunseen (mp, 0);	/* synchronize any Unseen-Sequence's      */
    seq_save (mp);		/* synchronize and save message sequences */
    folder_free (mp);		/* free folder/message structure          */

    context_save ();		/* save the global context file           */
    (void) m_unlink (tmpfilenam); /* remove temporary file                  */
    tmpfilenam = NULL;

    done (0);
    return 1;
}

/*
 * Clean up and exit
 */
static void NORETURN
unlink_done(int status)
{
    if (tmpfilenam && *tmpfilenam)
	(void) m_unlink (tmpfilenam);
    exit (status);
}
